src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -604,4 +618,223 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal interface IServerMutex : IDisposable
+    {
+        bool TryLock(int timeoutMs);
+        void Unlock();
+        bool IsDisposed { get; }
+    }
+
+    /// <summary>
+    /// An interprocess mutex abstraction based on OS advisory locking (FileStream.Lock/Unlock).
+    /// If multiple processes running as the same user create FileMutex instances with the same name,
+    ///  those instances will all point to the same file somewhere in a selected temporary directory.
+    /// The TryLock method can be used to attempt to acquire the mutex, with Unlock or Dispose used to release.
+    /// Unlike Win32 named mutexes, there is no mechanism for detecting an abandoned mutex. The file
+    ///  will simply revert to being unlocked but remain where it is.
+    /// </summary>
+    internal sealed class FileMutex : IDisposable
+    {
+        public readonly FileStream Stream;
+        public readonly string FilePath;
+
+        public bool IsLocked { get; private set; }
+
+        internal static string GetMutexDirectory()
+        {
+            var homeDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
```

Comment:
Mentioned `/tmp` before but I believe there are cygwin issues.

Reply:
Sorry, by that I meant `GetTempDirectory`, not literally `/tmp`

Reply:
Hm. Can we probe whether we're on NFS somehow, or probe whether advisory locks work? I'm guessing "questionable" means that probing wouldn't be reliable because the locks could work Sometimes

Reply:
Looks like probing whether a location is NFS is extremely not-trivial on *nix and requires installing an apt package on debian-based linux, so it's not reasonable to consider that.

Should I just use GetTempDirectory explicitly on mono and skip the home directories?

Reply:
I'd just call the `BuildServerConnection.GetTempPath` and use that until we find real cases where it breaks down.